### ISSUE TITLE
Video freezing with gnome 3.32 on wayland

### ISSUE THREAD

**Issue by jnsh** [mpv-player/mpv#6484]
### mpv version and platform
mpv-0.29.1-5, Arch linux

### Description of the issue
When watching videos on mpv with current gnome development branch 3.31 under wayland, and using `gpu-context=wayland` or `waylandvk`, the video output sometimes freezes while the audio keeps playing. The freeze can be triggered most of the time by pausing and resuming the playback, or by seeking the video. Also sometimes it occurs at seemingly random times.

The issue doesn't occur with gnome 3.30, so I bisected the [mutter commit](https://gitlab.gnome.org/GNOME/mutter/commit/2e79d05e04b0a124d20632ca0b3fcaa34f4aabce) that introduced this issue, and [reported](https://gitlab.gnome.org/GNOME/mutter/issues/457) it to gnome developers. However, according to their comments this would seem to be an issue with mpv's wayland backend.

This is how mutter developer Carlos Garnacho [explained](https://gitlab.gnome.org/GNOME/mutter/issues/457#note_431371) the commit that introduced the issue:
> Before that commit, wl_surface.commit would result on clutter_actor_set_scale() called when synchronizing buffer scale with the underlying actor. This would always trigger a full relayout and redraw, even though it's most often setting the same scale again and again.
>
> With that commit in place this does not happen anymore, which leaves surface/buffer damage as the way to trigger actor/buffer redraws (as it would be expected of the wayland protocol).

He then suggested running mpv with `WAYLAND_DEBUG=1`, to see whether surface/buffer damage happens when this issue is occuring, and that doesn't seem to be the case, suggesting that this would indeed be an mpv bug. Here's an excerpt from the logs:
```
# Here the image freezing bug has already occured, and mpv is currently paused
# press space to unpause
[3885451.702] wl_keyboard@19.key(2192, 26749438, 57, 1)
[3885452.337]  -> wl_surface@5.frame(new id wl_callback@28)
[3885452.427]  -> wl_surface@5.commit()
[3885552.174] wl_keyboard@19.key(2193, 26749550, 57, 0)
# The image stays frozen, while playback continues and audio still works
# press space to pause again
[3885912.011] wl_keyboard@19.key(2194, 26749910, 57, 1)
[3885914.997] wl_buffer@24.release()%) A-V:  0.000
[3885915.118] wl_callback@26.done(2138)
[3885916.898]  -> wl_surface@5.attach(wl_buffer@24, 0, 0)
[3885917.002]  -> wl_surface@5.damage(0, 0, 2147483647, 2147483647)
[3885917.062]  -> wl_surface@5.commit()
[3885917.082]  -> wl_display@1.sync(new id wl_callback@26)
[3885917.504] wl_display@1.delete_id(26)
[3885928.777] wl_display@1.delete_id(28)
[3886060.156] wl_keyboard@19.key(2195, 26750054, 57, 0)
# The image is refreshed once after pausing, but remains stuck if playback is un-paused again
```
Mutter developer @jadahl also [investigated](https://gitlab.gnome.org/GNOME/mutter/issues/457#note_431467) this issue, and this is his conclusion:
>To point things out, mpv seems to have two bugs:
>
>- It requests the frame callback after it has committed the surface state. It means there is a race where the content is drawn before the frame callback is queued, meaning the animation gets stuck seemingly out of the blue.
>
>- It doesn't kick start the animation when resuming, as it doesn't draw anything new, while still seemingly expecting the compositor to composite a new frame, replying to the frame callback.

See the aforementioned mutter issue for more details.

Additionally, I think this is the same/related bug encountered in #5914. As with that bug, this issue only seems to happen after https://github.com/mpv-player/mpv/commit/a743fef837bcab206b1e576db7e7a64b02890449.

**Comment by jnsh** [comment]
@atomnuker Could you by any chance have a look at this, since you authored the commit that results in this issue.

**Comment by atomnuker** [comment]
Well they're wrong and the rest of compositors agree.

**Comment by emersion** [comment]
Ref https://gitlab.freedesktop.org/xorg/xserver/merge_requests/116#note_111464

tl;dr sending `wl_surface.frame` followed by `wl_surface.commit` is correct behaviour.

**Comment by jnsh** [comment]
Thank you both for the responses.
I reported this back to the mutter issue I mentioned on the first post.

**Comment by X7SHaDe** [comment]
Confirmed also happening with KDE/Plasma (wayland session). The video hangs.

**Comment by emersion** [comment]
Please fill a KDE bug.

**Comment by jnsh** [comment]
The issue has now been fixed in GNOME/mutter git master with [this commit](https://gitlab.gnome.org/GNOME/mutter/commit/d49d10b14f4e0fa80e6867979b26fab383610b39). I'm assuming the fix will be included in mutter-3.34.2.
